.circleci/config.yml

```diff
@@ -13,101 +13,122 @@ jobs:
       - save_cache:
           paths:
-            - ~/.local
-            - ~/pixel/.venv
-          key: backend-deps-v3-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
+            - ~/.local/share/virtualenvs
+          key: backend-dev-deps-v1-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
 
   lint_backend:
-    working_directory: ~/pixel
+    working_directory: /app/pixel
     docker:
-      - image: circleci/python:3.6.1
-        environment:
-          PIPENV_VENV_IN_PROJECT: 1
+      - image: candihub/pixel:$CIRCLE_BRANCH
     steps:
-      - checkout
       - restore_cache:
           keys:
-          - backend-deps-v3-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
-      - run:
-          name: Add Pipenv & installed tools to the PATH
-          command: |
-            echo "export PATH=~/.local/bin:~/pixel/.venv/bin:$PATH" >> $BASH_ENV
+            - backend-dev-deps-v1-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
       - run:
           name: Lint the code
-          command: make lint
+          command: |
+            make lint-ci
 
   test_backend:
-    working_directory: ~/pixel
+    working_directory: /app/pixel
     docker:
-      - image: circleci/python:3.6.1
+      - image: candihub/pixel:$CIRCLE_BRANCH
+        environment:
+          DJANGO_SETTINGS_MODULE: pixel.settings
+          DJANGO_CONFIGURATION: Test
+          DJANGO_SECRET_KEY: ThisIsAnExampleKeyForDevPurposeOnly
+          POSTGRES_HOST: localhost
+          POSTGRES_DB: pixel
+          POSTGRES_USER: pixel
+          POSTGRES_PASSWORD: pixel42
+      - image: postgres:9.6
         environment:
-          PIPENV_VENV_IN_PROJECT: 1
+          POSTGRES_DB: pixel
+          POSTGRES_USER: pixel
+          POSTGRES_PASSWORD: pixel42
     steps:
-      - checkout
       - restore_cache:
           keys:
-          - backend-deps-v3-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
+            - backend-dev-deps-v1-{{ .Branch }}-{{ checksum "Pipfile.lock" }}
       - run:
-          name: Add Pipenv & installed tools to the PATH
+          name: Test the code
           command: |
-            echo "export PATH=~/.local/bin:~/pixel/.venv/bin:$PATH" >> $BASH_ENV
+            sleep 10
+            make test-ci
       - run:
-          name: Test the code
+          name: Create git repository
           command: |
-            make test
+            git init
+            git remote add origin $CIRCLE_REPOSITORY_URL
+            mkdir ~/.ssh
+            ssh-keyscan -t rsa github.com >> ~/.ssh/known_hosts
+            git fetch
+            git reset origin/$CIRCLE_BRANCH
+            git checkout origin/$CIRCLE_BRANCH
       - run:
           name: Report code coverage
           command: |
-            make coverage
+            make coverage-ci
           when: on_success
 
 workflows:
   version: 2
 
   pixel:
     jobs:
-      - build_backend
       - build_frontend
-      - lint_backend:
+      - build_backend:
+          requires:
+            - build_frontend
```

Comment:
Given that we will not bundle frontend and backend separately, I would rather add the "build frontend" step in the build step. There are two many jobs here, and it implies two things:

1. the circle-ci UI is completely unreadable
2. it takes more time to complete a CI build

Reply:
🤔 not sure to fully agree with you on this. I mean, I agree that there are many steps, but it stays readable and the front-end build job take a few seconds (thanks to the cache).

Reply:
Given that we will not bundle frontend and backend separately, I really dont see the point of separating both.

Reply:
Not mixing building environments. If you build front and back in the same job, you will need both node and Docker environments in the same machine or container (I don't know if the second is allowed in CircleCI).